docs/tendermint-core/p2p/peer_manager.md

```diff
@@ -0,0 +1,439 @@
+The peer manager distinguishes *incoming* from *outgoing* connections.
+A connection is *incoming* when the router has [accepted](#accepted-transition)
+it from a peer.
+A connection is *outgoing* when the router has successfully
+[dialed](#dialed-transition) a peer.
+
+If the `MaxOutgoingConnections` parameter is set (it is larger than zero), it
+defines the maximum number of *outgoing* connections the node should maintain.
+More precisely, it determines that the node should not attempt to dial new
+peers when the router already has established outgoing connections to
+`MaxOutgoingConnections` peers.
+
+> The previous version of the `p2p` explicitly distinguished incoming and
+> outgoing peers. Configuring the `MaxOutgoingConnections` parameters should
+> therefore make the connection policy similar to the one adopted in the
+> previous version. (TODO: check)
+
+### Slot upgrades
+
+The rationale behind this concept is that the node may try to establish or
+accept connections even when all the connection slots are full, provided that
+the peer in the other side of the new connection is better-ranked than a peer
+that is occupying a connection slot.
+A slot can therefore be upgraded, meaning that the lower-ranked peer
+occupying the slot will be replaced by a higher-ranked peer.
+
+The upgrading of connection slots is determined by the `MaxConnectedUpgrade`
+parameter, which defines the number of connections that the peer manager can
+use for upgrading connection slots.
+
+If `MaxConnectedUpgrade` is set to zero, the upgrading of connection slots is
+disabled.
+This means, in particular, that `MaxConnected` is the hard limit of peers that
+can be in the [connected state](#connected-peer).
+
+If `MaxConnectedUpgrade` is larger than zero, the upgrading of connection slots
+is enabled.
+As a result, the hard limit for the number of peers that can be in the
+[connected state](#connected-peer) becomes `MaxConnected + MaxConnectedUpgrade`.
+Some of these peers, however, will not remain in this state as they should be
+[evicted](#evictnext-transition) by the router.
+
+### Peer ranking
+
+The peer manager should rank peers based on user-provided parameters and on the
+current state of the peer.
+
+The ranking is established by ordering all known peers by its score.
+This mechanism is currently very basic.
+
+> The code contains a number of potential replacements for this ranking
+> mechanism. Therefore, improving this mechanism is a work in progress.
+
+Peers configured as `PersistentPeers` have _always_ `PeerScorePersistent`,
+which is the maximum allowed peer score.
+
+The remaining peers have a `MutableScore`, initialized to `0` when the peer is
+added to the peer manager.
+When the peer is reported as a `PeerStatusGood`, its score is incremented.
+When the peer is reported as a `PeerStatusBad`, its score is decremented.
+
+> The mechanisms based on the "reputation" of the peer according to reactors,
+> however, appears not to be fully implemented.
```

Comment:
We do not understand this effect.

Currently, we only punish peers reported as "bad" (when a reactor reported an error) and temporarily decrease the score of peers to which we fail to connect. Notice that we only decrease the score of a peer, which means that it will not become a higher-ranked peer with these procedures.

A good question I just thought about: when we connect to a peer that we failed to dial before, we reset the dial fail counters. This means that the penalty on score of this peer are removed, meaning that its score increases. I think this is done *after*  we decided or not to accept the peer, so its new score is not considered at this point. But the effect of this actual variation of peer score (as this is the only one actually implemented) should be better analyzed.

Reply:
I think we're on the same page on this